Tests/AK/TestStatistics.cpp

```diff
@@ -12,6 +12,8 @@ TEST_CASE(Statistics)
     // Setup Test Data
     AK::Statistics<double> odd_number_elements;
     AK::Statistics<double> even_number_elements;
+    AK::Statistics<double> odd_number_elements_201;
+    AK::Statistics<double> even_number_elements_350;
```

Comment:
Can this go into a separate test case? Ideally, add something like `EXPECT(201 > Statistics::EVEN_NAIVE_MEDIAN_CUTOFF); EXPECT(350 > Statistics::ODD_NAIVE_MEDIAN_CUTOFF);`, so that the test fails when those numbers change.

Also, I believe this would make the condition `(size() <= EVEN_NAIVE_MEDIAN_CUTOFF && size() % 2 == 0)` true, doesn't it? In that case it would still fail to cover the `quick_select` invocation.

Reply:
Makes sense. I will do this.

The quick select invocations will be covered by both of of those. I explicitly checked using debug prints i obiously didn't commit. The two 6 element versions will deal with the naive version whilst the two big ones will deal with both quickselect cases.

Reply:
Awesome! Can you do me a favor and walk me through this? I really don't understand what's going on then.

We initialize the dataset to have 350 elements. so `size()` will return 350. `EVEN_NAIVE_MEDIAN_CUTOFF` is defined in `Statistics.h` to be 350. So I would expect that the `(size() <= EVEN_NAIVE_MEDIAN_CUTOFF && size() % 2 == 0)` branch is taken. You claim that something different happens. You're probably right, but could you explain why?

Reply:
No I was actually thinking you meant that branch being taken, since that's what I am testing here.

The who cases with large datasets are supposed to be testing the even and odd quick select cases respectively.
The Two short ones with either 5 or 6 elements are meant to be testing the naive implementation of median.

I think I'm missunderstanding you're original question?

Reply:
Ah, I never spelled out the actual question I have, my bad, sorry.

Let's look at the invocation `even_number_elements_350.median()`.
- Do I read that right that the *intention* is that this covers the "large dataset, even quick select" case?
- With the above reasoning, I would strongly expect that the "SMALL dataset, even quickSORT" branch is taken.

I added a few prints, just to see where me logic goes wrong: https://github.com/BenWiederhake/serenity/commit/7bfbecf86c89e9e518214685f84dfc0d97bfd1e9

This is the output:
```
$ ./TestStatistics 
Running 1 cases out of 1.
Running test 'Statistics'.
Running TestStatistics median, small odd
Statistics::median(): size()=5, odd quickSORT taken
Running TestStatistics median, small even
Statistics::median(): size()=6, even quickSORT taken
Running TestStatistics median, large odd
Statistics::median(): size()=201, odd quickSELECT taken
Running TestStatistics median, large even
Statistics::median(): size()=350, even quickSORT taken
Finished TestStatistics median
Completed test 'Statistics' in 0ms
Finished 1 tests and 0 benchmarks in 0ms (0ms tests, 0ms benchmarks, 0ms other).
Out of 1 tests, 1 passed and 0 failed.
```

As you can see, the even quickSELECT branch is not covered by this test, even though it tries to do so. Let's improve that, by:
- changing "350" to "360",
- replacing the numbers in the variable names by "large", and
- adding a `VERIFY()` that checks that the hard-coded numbers definitely are larger than the respective threshold.

Reply:
Thank you, I'll do that once i got time.